azblob/zc_retry_reader.go

```diff
@@ -45,19 +64,37 @@ type RetryReaderOptions struct {
 // through reading from the new response.
 type retryReader struct {
 	ctx             context.Context
-	response        *http.Response
 	info            HTTPGetterInfo
 	countWasBounded bool
 	o               RetryReaderOptions
 	getter          HTTPGetter
+
+	// we support Close-ing during Reads (from other goroutines), so we protect the shared state, which is response
+	responseMu       *sync.Mutex
+	response         *http.Response
 }
 
 // NewRetryReader creates a retry reader.
 func NewRetryReader(ctx context.Context, initialResponse *http.Response,
 	info HTTPGetterInfo, o RetryReaderOptions, getter HTTPGetter) io.ReadCloser {
-	return &retryReader{ctx: ctx, getter: getter, info: info, countWasBounded: info.Count != CountToEnd, response: initialResponse, o: o}
+	return &retryReader{
+		ctx: ctx,
+		getter: getter,
+		info: info,
+		countWasBounded: info.Count != CountToEnd,
+		response:    initialResponse,
+		responseMu:  &sync.Mutex{},
+		o:           o}
 }
 
+
+func (s *retryReader) setResponse(r *http.Response){
+	s.responseMu.Lock()
+	defer s.responseMu.Unlock()
```

Comment:
Seems defer is not necessary here. What about use following to save per cost brought by defer?

s.responseMu.Lock()
s.response = r
s.responseMu.Unlock()

Reply:
Could do it that way.  Defer seems nice and clean to read and perf cost is negligible compared to what the code is actually doing here (i.e. its just done a new HTTP request).  The cost of a defer will be many, many orders of magnitude smaller, so I'm inclined to leave this as is and consider this resolved.  Does that sound OK?